x-pack/plugins/uptime/public/components/monitor/ping_list/columns/ping_timestamp/step_image_popover.tsx

```diff
@@ -24,40 +26,108 @@ const StepImage = styled(EuiImage)`
     }
   }
 `;
+
+interface ScreenshotImageProps {
+  captionContent: string;
+  imageCaption: JSX.Element;
+}
+
+const DefaultImage: React.FC<ScreenshotImageProps & { url?: string }> = ({
+  captionContent,
+  imageCaption,
+  url,
+}) =>
+  url ? (
+    <StepImage
+      allowFullScreen={true}
+      alt={captionContent}
+      caption={imageCaption}
+      data-test-subj="pingTimestampImage"
+      hasShadow
+      url={url}
+      size="s"
+      className="syntheticsStepImage"
+    />
+  ) : (
+    <EuiLoadingSpinner size="l" />
+  );
+
+/**
+ * This component provides an intermediate step for composite images. It causes a loading spinner to appear
+ * while the image is being re-assembled, then calls the default image component and provides a data URL for the image.
+ */
+const RecomposedScreenshotImage: React.FC<
+  ScreenshotImageProps & {
+    imgRef: ScreenshotRefImageData;
+    setUrl: React.Dispatch<string | undefined>;
+    url?: string;
+  }
+> = (props) => {
+  const { imgRef, setUrl, url } = props;
+
+  // initially an undefined URL value is passed to the image display, and a loading spinner is rendered.
+  // `useCompositeImage` will call `setUrl` when the image is composited, and the updated `url` will display.
+  useCompositeImage(imgRef, setUrl, url);
+
+  return <DefaultImage {...props} url={props.url} />;
+};
+
 export interface StepImagePopoverProps {
   captionContent: string;
   imageCaption: JSX.Element;
-  imgSrc: string;
+  imgSrc?: string;
+  imgRef?: ScreenshotRefImageData;
   isImagePopoverOpen: boolean;
 }
 
+const StepImageComponent: React.FC<
+  Omit<StepImagePopoverProps, 'isImagePopoverOpen'> & {
+    setUrl: React.Dispatch<string | undefined>;
+    url: string | undefined;
+  }
+> = (props) => {
+  if (props.imgSrc) {
+    props.setUrl(props.imgSrc);
+    return <DefaultImage {...props} url={props.url} />;
+  } else if (props.imgRef) {
+    return <RecomposedScreenshotImage {...props} imgRef={props.imgRef} />;
+  }
+  return null;
+};
+
 export const StepImagePopover: React.FC<StepImagePopoverProps> = ({
   captionContent,
   imageCaption,
+  imgRef,
   imgSrc,
   isImagePopoverOpen,
-}) => (
-  <EuiPopover
-    anchorPosition="leftDown"
-    button={
-      <StepImage
-        allowFullScreen={true}
-        alt={captionContent}
-        caption={imageCaption}
-        data-test-subj="pingTimestampImage"
-        hasShadow
-        url={imgSrc}
-        size="s"
-        className="syntheticsStepImage"
-      />
-    }
-    isOpen={isImagePopoverOpen}
-    closePopover={() => {}}
-  >
-    <EuiImage
-      alt={fullSizeImageAlt}
-      url={imgSrc}
-      style={{ height: POPOVER_IMG_HEIGHT, width: POPOVER_IMG_WIDTH, objectFit: 'contain' }}
-    />
-  </EuiPopover>
-);
+}) => {
+  const [url, setUrl] = React.useState<string | undefined>(undefined);
```

Comment:
Yeah this is a good point; there are three logical paths here:

1. `imgSrc` is defined
2. `imgRef` is defined
3. neither is defined

If `imgSrc`, it's a Screenshot blob, so we can just go ahead and render it.

Otherwise if `imgRef`, it's a ref that we need to compose.

In either case, we're going to end up with a string, which we end up storing in this `url` field, which we should probably rename to `[imageData, setImageData]`.

I think when I wrote this, I was thinking it'd be nicer to encapsulate these logical paths in a dedicated component as an implementation detail above, but if it's unclear I could refactor it to be inline here.

Reply:
++ to Dominique comment